parser.js

```diff
@@ -20,16 +20,27 @@ const visitorKeys = require("./visitor-keys");
 
 exports.version = require("./package.json").version;
 
-exports.parseForESLint = function parseForESLint(code, options) {
-    if (typeof options !== "object" || options === null) {
-        options = { useJSXTextNode: true };
-    } else if (typeof options.useJSXTextNode !== "boolean") {
-        options = Object.assign({}, options, { useJSXTextNode: true });
+exports.parseForESLint = function parseForESLint(code, inputOptions) {
+    const options = Object.assign({
+        useJSXTextNode: true,
+        // typescript-estree doesn't respect ecmaFeatures object
+        jsx: false
+    }, inputOptions);
+
+    if (typeof options.useJSXTextNode !== "boolean") {
+        options.useJSXTextNode = true;
     }
-    if (typeof options.filePath === "string") {
-        const tsx = options.filePath.endsWith(".tsx");
-        if (tsx || options.filePath.endsWith(".ts")) {
-            options = Object.assign({}, options, { jsx: tsx });
+
+    options.jsx = (options.ecmaFeatures || {}).jsx;
+    if (typeof options.jsx !== "boolean") {
+        options.jsx = false;
+    }
+
+    // override the jsx option depending on the file path
+    if (typeof inputOptions.filePath === "string") {
```

Comment:
I'm not convinced this is the behavior we want. Since the parser currently sets this automatically based on file extension, this option should only be necessary when no file path is present. As such, autodetection should not happen when this option is set.

Reply:
there is few examples of cases with no filePath:

vue files provided by `vue-eslint-parser`,
files provided by `eslint-html-parser`, `eslint-plugin-html`, 
and potentially files provided by https://github.com/eslint/eslint-plugin-markdown in feature..

Reply:
I was just going by the behaviour listed in the readme:

- It's `false` on `*.ts` files _regardless of this option_.
- It's `true` on `*.tsx` files _regardless of this option_.

Reply:
yes but this is done by typescript itself https://www.typescriptlang.org/docs/handbook/jsx.html

> In order to use JSX you must do two things.
>
> * Name your files with a .tsx extension
> * Enable the jsx option

Reply:
Sorry, I didn’t realize this was the current behavior.